src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -239,4 +242,107 @@ std::shared_ptr<TableStatistics> ChunkPruningRule::_prune_table_statistics(const
                                            old_statistics.row_count - static_cast<float>(num_rows_pruned));
 }
 
+std::vector<PredicateChain> ChunkPruningRule::_find_predicate_chains_recursively(
+    const std::shared_ptr<StoredTableNode>& stored_table_node, const std::shared_ptr<AbstractLQPNode>& node,
+    PredicateChain current_predicate_chain) {
+  std::vector<PredicateChain> predicate_chains;
+
+  visit_lqp_upwards(node, [&](const auto& current_node) {
+    /**
+     * In the following switch-statement, we
+     *  (1) add PredicateNodes to the current predicate chain, if applicable, and
+     *  (2) check whether the predicate chain continues or ends with current_node.
+     */
+    auto predicate_chain_continues = true;
+    switch (current_node->type) {
+      // clang-format off
+      case LQPNodeType::Alias:
+      case LQPNodeType::Sort:
+      case LQPNodeType::StoredTable:
+      case LQPNodeType::Validate:
+      case LQPNodeType::Projection:   break;
+      case LQPNodeType::Predicate: {
+        const auto& predicate_node = std::static_pointer_cast<PredicateNode>(current_node);
+
+        // PredicateNode might not belong to the current predicate chain,
+        // e.g. when it follows a JoinNode and contains LQPColumnExpressions from other StoredTableNodes.
```

Comment:
Another thing to add to the clarification of what a chain is: It appears as if these chains are broken by joins in some rules and continue across certain types of joins in other nodes.

Also, would this mean that two chains might "overlap", i.e., one chain starting on the left side of a join, one on the right, and both continuing above?

Reply:
> Also, would this mean that two chains might "overlap", i.e., one chain starting on the left side of a join, one on the right, and both continuing above?

Indeed. Therefore, we might visit nodes multiple times.

Reply:
I had my second thought on this @mrks 
Since Hyrise does not support self-joins that originate from a single StoredTableNode, predicate pruning chains should never merge after having branched. 
I added a `visited_nodes` data structure and an Assert that fails whenever we see nodes twice. The [CI has passed](https://hyrise-ci.epic-hpi.de/blue/organizations/jenkins/hyrise%2Fhyrise/detail/PR-2284/43/pipeline/), so I guess we are fine.

Reply:
Who says that we need a self-join for this? Wouldn't this also occur if we have a disjunctive predicate that gets split up into a diamond?

Reply:
True. But, since we cancel predicate pruning chains with the occurrence of Union nodes anyway, we do not run into problems, do we?

Reply:
I see. Where did the assert go though?

Reply:
```cpp
std::vector<PredicatePruningChain> find_predicate_pruning_chains_by_stored_table_node_recursively(
    const std::shared_ptr<AbstractLQPNode>& next_node, PredicatePruningChain current_predicate_pruning_chain,
    const std::shared_ptr<StoredTableNode>& stored_table_node,
    std::unordered_set<std::shared_ptr<AbstractLQPNode>> visited_nodes) {
  std::vector<PredicatePruningChain> predicate_pruning_chains;

  visit_lqp_upwards(next_node, [&](const auto& current_node) {
    Assert(!visited_nodes.contains(current_node),
           "Did not expect to see the same node twice.");

...
```

Reply:
Github... :(

Reply:
> Github... :(

Interesting, the duplicated comments have disappeared. 😄